lib/msal-browser/src/interaction_handler/SilentHandler.ts

```diff
@@ -0,0 +1,175 @@
+import { BrowserStorage } from "../cache/BrowserStorage";
+
+export class SilentHandler extends InteractionHandler {
+
+    private loadFrameTimeout: number;
+    constructor(authCodeModule: AuthorizationCodeModule, storageImpl: BrowserStorage, configuredLoadFrameTimeout: number) {
+        super(authCodeModule, storageImpl);
+        this.loadFrameTimeout = configuredLoadFrameTimeout;
+    }
+
+    /**
+     * Creates a hidden iframe to given URL.
+     * @param urlNavigate 
+     */
+    async initiateAuthRequest(requestUrl: string): Promise<HTMLIFrameElement> {
+        const frameName = "msalIdTokenFrame";
+        return this.loadFrameTimeout ? await this.loadFrame(requestUrl, frameName) : this.loadFrameSync(requestUrl, frameName);
+    }
+
+    /**
+     * Monitors an iframe content window until it loads a url with a known hash, or hits a specified timeout.
+     * @param iframeContentWindow 
+     * @param timeout 
+     * @param urlNavigate 
+     */
+    monitorFrameForHash(iframe: HTMLIFrameElement, timeout: number, urlNavigate: string): Promise<string> {
+        const contentWindow = iframe.contentWindow;
+        return new Promise((resolve, reject) => {
+            const maxTicks = timeout / BrowserConstants.POLL_INTERVAL_MS;
+            let ticks = 0;
+
+            const intervalId = setInterval(() => {
+                if (contentWindow.closed) {
+                    // Window is closed
+                    this.removeHiddenIframe(iframe);
+                    clearInterval(intervalId);
+                    reject(BrowserAuthError.createIframeClosedPrematurelyError());
+                    return;
+                }
+
+                let href;
+                try {
+                    /*
+                     * Will throw if cross origin,
+                     * which should be caught and ignored
+                     * since we need the interval to keep running while on STS UI.
+                     */
+                    href = contentWindow.location.href;
+                } catch (e) {}
+
+                /*
+                 * Always run clock for silent calls
+                 * as silent operations should be short,
+                 * and to ensure they always at worst timeout.
+                 */
+                ticks++;
+
+                if (UrlString.hashContainsKnownProperties(href)) {
+                    // Success case
+                    const contentHash = contentWindow.location.hash;
+                    this.removeHiddenIframe(iframe);
+                    clearInterval(intervalId);
+                    resolve(contentHash);
+                    return;
+                } else if (ticks > maxTicks) {
+                    // Timeout error
+                    this.removeHiddenIframe(iframe);
+                    clearInterval(intervalId);
+                    reject(BrowserAuthError.createMonitorWindowTimeoutError(urlNavigate));
+                    return;
```

Comment:
Nitpick: `return` isnt needed

Reply:
I'd like to keep it here to make sure we are ending the function after rejecting, this has caused problems in the past.